File: ImperatorToCK3/Source/CommonUtilities/DatedHistoryBlock.cpp
#include "DatedHistoryBlock.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



DatedHistoryBlock::DatedHistoryBlock(std::vector<SimpleFieldStruct> _simpleFieldStructs, std::vector<ContainerFieldStruct> _containerFieldStructs, std::istream& theStream):
	simpleFieldStructs(std::move(_simpleFieldStructs)), containerFieldStructs(std::move(_containerFieldStructs))
{
	for (const auto& [fieldName, setter, _initialValue] : simpleFieldStructs) {
		registerKeyword(setter, [&](std::istream& theStream) {
			contents.simpleFieldContents[fieldName].emplace_back(commonItems::getString(theStream));
		});
	}
	for (const auto& [fieldName, setter, _initialValue] : containerFieldStructs) {
		registerKeyword(setter, [&](std::istream& theStream) {
			contents.containerFieldContents[fieldName].emplace_back(commonItems::getStrings(theStream));
		});
	}
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
	parseStream(theStream);
	clearRegisteredKeywords();
}

File: ImperatorToCK3/CommonUtils/DatedHistoryBlock.cs
using commonItems;
using System.Collections.Generic;
using System.Linq;

namespace ImperatorToCK3.CommonUtils;

public class DatedHistoryBlock : Parser {
	public Dictionary<string, List<FieldValue>> SimpleFieldContents { get; } = new();
	public Dictionary<string, List<FieldValue>> ContainerFieldContents { get; } = new();

	public DatedHistoryBlock(IEnumerable<SimpleFieldDef> simpleFieldStructs, IEnumerable<ContainerFieldDef> containerFieldStructs, BufferedReader reader) {
		foreach (var fieldStruct in simpleFieldStructs) {
			foreach (var setter in fieldStruct.Setters) {
				RegisterKeyword(setter, reader => {
					if (!SimpleFieldContents.ContainsKey(fieldStruct.FieldName)) {
						SimpleFieldContents.Add(fieldStruct.FieldName, new());
					}
					SimpleFieldContents[fieldStruct.FieldName].Add(new FieldValue(HistoryFactory.GetValue(reader.GetString()), setter));
				});
			}
		}
		foreach (var fieldStruct in containerFieldStructs) {
			foreach (var setter in fieldStruct.Setters) {
				RegisterKeyword(setter, reader => {
					if (!ContainerFieldContents.ContainsKey(fieldStruct.FieldName)) {
						ContainerFieldContents.Add(fieldStruct.FieldName, new());
					}

					var strings = reader.GetStrings();
					var values = new List<object>(strings.Select(HistoryFactory.GetValue));
					ContainerFieldContents[fieldStruct.FieldName].Add(new FieldValue(values, setter));
				});
			}
		}
		RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
		ParseStream(reader);
		ClearRegisteredRules();
	}
}
